sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -69,18 +69,58 @@ type ctxKey string
 const bundleKey ctxKey = "beam:bundle"
 const ptransformKey ctxKey = "beam:ptransform"
 
+// context is a caching context since reads & writes are expensive.
+type beamCtx struct {
+	context.Context
+	bundleID, ptransformID string
+}
+
+// Lift the keys value for faster lookups when not available.
+func (ctx *beamCtx) Value(key interface{}) interface{} {
+	switch key {
+	case bundleKey:
+		if ctx.bundleID == "" {
+			if id := ctx.Value(key); id != nil {
+				ctx.bundleID = id.(string)
+			}
+		}
+		return ctx.bundleID
+	case ptransformKey:
+		if ctx.ptransformID == "" {
+			if id := ctx.Value(key); id != nil {
+				ctx.ptransformID = id.(string)
+			}
+		}
+		return ctx.ptransformID
+	}
+	return ctx.Context.Value(key)
+}
+
 // SetBundleID sets the id of the current Bundle.
 func SetBundleID(ctx context.Context, id string) context.Context {
-	return context.WithValue(ctx, bundleKey, id)
+	var ptransformID string
+	if bctx, ok := ctx.(*beamCtx); ok {
+		ptransformID = bctx.ptransformID
+	}
+	return &beamCtx{Context: ctx, bundleID: id, ptransformID: ptransformID}
```

Comment:
We can also use the beamCtx's context as Context here to avoid nesting beamCtxs.

Reply:
I considered that but then metrics following an emit would be incorrect, short of retaining the parent PTransform ID and re-adding that, which would work, but would be cumbersome. That approach is a viable alternative to caching the Context in pardo.go though as it accomplishes the same goal. 
Do you prefer that approach?

eg. re: correctness issue on ordering of event if the context isn't re-allocated:
SDK sets PtransformID
SDK calls User ParDo
User ParDo1 emits - > SDK runs child ParDos which set their own PTransformIDs
User ParDo1 logs a metric -> Metric is associated with the wrong PTransform.

That is, we'd need to extract and re-assert the parent's PTransform ID before returning from the ParDo process elements.

Reply:
I was unclear. I was suggesting a simple improvement to these helpers avoid nesting beamCtxs by adding a sibling to the input chain of a beamCtx:

func SetBundleID(ctx context.Context, id string) context.Context {
	if bctx, ok := ctx.(*beamCtx); ok {
		return &beamCtx{Context: bctx.ctx, bundleID: id, ptransformID: bctx.ptransformID}
	}
	return &beamCtx{Context: ctx, bundleID: id}
}